src/mjolnir/linkclassification.cc

```diff
@@ -134,272 +140,466 @@ nodelist_t FormExitNodes(sequence<Node>& nodes, sequence<Edge>& edges) {
-std::pair<uint32_t, uint32_t> Reclassify(LinkTreeNode& root,
-                                         sequence<Edge>& edges,
-                                         sequence<OSMWay>& ways,
-                                         sequence<OSMWayNode>& way_nodes,
-                                         std::queue<LinkTreeNode*>& leaves,
-                                         bool infer_turn_channels) {
-  // Get the classification at the root node.
-  std::pair<uint32_t, uint32_t> counts = std::make_pair(0, 0);
-  uint32_t exit_classification = root.classification;
-
-  // Work up from each leaf until the root is found or a branch occurs.
-  // Add branches to expand to a list;
-  uint32_t rc = 0;
-  std::vector<uint32_t> link_edge_indexes;
-  while (!leaves.empty()) {
-    // Clear the list of link edge indices from the prior leaf
-    link_edge_indexes.clear();
+// Helper structure that contains all osm data we need in one place
+struct Data {
+  Data(const std::string& nodes_file,
+       const std::string& edges_file,
+       const std::string& ways_file,
+       const std::string& way_nodes_file,
+       const OSMData& osmdata)
+      : nodes(nodes_file, false), edges(edges_file, false), ways(ways_file, false),
+        way_nodes(way_nodes_file, false), osmdata(osmdata) {
+  }
 
-    // Get the next leaf node and its classification
-    LinkTreeNode* leaf_node = leaves.front();
-    LinkTreeNode* current_node = leaf_node;
-    leaves.pop();
+  sequence<Node> nodes;
+  sequence<Edge> edges;
+  sequence<OSMWay> ways;
+  sequence<OSMWayNode> way_nodes;
+  const OSMData& osmdata;
+};
 
-    // Track information required for turn channel tests.
-    bool has_fork = false;
-    bool has_exit = current_node->has_exit;
-    bool ends_have_non_link = current_node->non_link_count > 0;
+// Way tags that are used to determine correct link class
+struct WayTags {
+  std::vector<std::string> refs;
+  std::vector<std::string> dest_refs;
 
-    // Go upward in the tree until a branch is found or the root is reached
-    while (current_node->parent != nullptr) {
-      LinkTreeNode* parent = current_node->parent;
-      if (parent->link_count > 2) {
-        has_fork = true;
-      }
-      if (parent->has_exit) {
-        has_exit = true;
-      }
+  bool is_empty() const {
+    return refs.empty() && dest_refs.empty();
+  }
 
-      // Get the link edge index from the parent to the current node
-      for (uint32_t i = 0; i < parent->children.size(); ++i) {
-        if (&parent->children[i] == current_node) {
-          link_edge_indexes.push_back(parent->link_edge_index[i]);
-          break;
-        }
-      }
+  static WayTags Parse(const OSMWay& way, const OSMData& osmdata) {
+    WayTags road_tags;
+
+    // Parse 'destination:ref' tag
+    if (way.destination_ref_index() != 0)
+      road_tags.dest_refs = GetTagTokens(osmdata.name_offset_map.name(way.destination_ref_index()));
+
+    // Parse 'ref' tag
+    if (way.ref_index() != 0)
```

Comment:
@genadz should we be looking at `int_ref` too?

Reply:
according to the https://taginfo.openstreetmap.org/tags/highway=motorway#combinations , `ref` tags is used in more than `94%` cases while `int_ref` is used only in `~23%`. For trunk and primary roads we have similar picture (`ref` tag is much more popular). So, I think `int_ref` tag won't affect classification too much.

and , in general , we definitely can use other tags (like `destination:street; destination:ref:lanes; destination:ref:forward/backward`), but all of these require additional testing round. I'd propose in this PR focus only on `destination:ref; ref` tags